bin/setclasspath.sh

```diff
@@ -0,0 +1,105 @@
+#  Set JAVA_HOME or JRE_HOME if not already set, ensure any provided settings
+#  are valid and consistent with the selected start-up options and set up the
+#  endorsed directory.
+# -----------------------------------------------------------------------------
+
+# Make sure prerequisite environment variables are set
+if [ -z "$JAVA_HOME" ] && [ -z "$JRE_HOME" ]; then
+  if $darwin; then
+    # Bugzilla 54390
+    if [ -x '/usr/libexec/java_home' ] ; then
+      export JAVA_HOME=`/usr/libexec/java_home`
+    # Bugzilla 37284 (reviewed).
+    elif [ -d "/System/Library/Frameworks/JavaVM.framework/Versions/CurrentJDK/Home" ]; then
+      export JAVA_HOME="/System/Library/Frameworks/JavaVM.framework/Versions/CurrentJDK/Home"
+    fi
+  else
+    JAVA_PATH=`which java 2>/dev/null`
+    if [ "x$JAVA_PATH" != "x" ]; then
+      JAVA_PATH=`dirname "$JAVA_PATH" 2>/dev/null`
+      JRE_HOME=`dirname "$JAVA_PATH" 2>/dev/null`
+    fi
+    if [ "x$JRE_HOME" = "x" ]; then
+      # XXX: Should we try other locations?
+      if [ -x /usr/bin/java ]; then
+        JRE_HOME=/usr
+      fi
+    fi
+  fi
+  if [ -z "$JAVA_HOME" ] && [ -z "$JRE_HOME" ]; then
+    echo "Neither the JAVA_HOME nor the JRE_HOME environment variable is defined"
+    echo "At least one of these environment variable is needed to run this program"
+    exit 1
+  fi
+fi
+if [ -z "$JAVA_HOME" ] && [ "$1" = "debug" ]; then
+  echo "JAVA_HOME should point to a JDK in order to run in debug mode."
+  exit 1
+fi
+if [ -z "$JRE_HOME" ]; then
+  JRE_HOME="$JAVA_HOME"
+fi
+
+# If we're running under jdb, we need a full jdk.
+if [ "$1" = "debug" ] ; then
+  if [ "$os400" = "true" ]; then
+    if [ ! -x "$JAVA_HOME"/bin/java ] || [ ! -x "$JAVA_HOME"/bin/javac ]; then
+      echo "The JAVA_HOME environment variable is not defined correctly"
+      echo "This environment variable is needed to run this program"
+      echo "NB: JAVA_HOME should point to a JDK not a JRE"
+      exit 1
+    fi
+  else
+    if [ ! -x "$JAVA_HOME"/bin/java ] || [ ! -x "$JAVA_HOME"/bin/jdb ] || [ ! -x "$JAVA_HOME"/bin/javac ]; then
+      echo "The JAVA_HOME environment variable is not defined correctly"
+      echo "This environment variable is needed to run this program"
+      echo "NB: JAVA_HOME should point to a JDK not a JRE"
+      exit 1
+    fi
+  fi
+fi
+
+# Don't override the endorsed dir if the user has set it previously
+if [ -z "$JAVA_ENDORSED_DIRS" ]; then
+  # Java 9 no longer supports the java.endorsed.dirs
+  # system property. Only try to use it if
+  # CATALINA_HOME/endorsed exists.
+  if [ -d "$CATALINA_HOME"/endorsed ]; then
+    JAVA_ENDORSED_DIRS="$CATALINA_HOME"/endorsed
+  fi
+fi
+
+# Set standard commands for invoking Java, if not already set.
+if [ -z "$_RUNJAVA" ]; then
+  _RUNJAVA="$JRE_HOME"/bin/java
+fi
+if [ "$os400" != "true" ]; then
+  if [ -z "$_RUNJDB" ]; then
+    _RUNJDB="$JAVA_HOME"/bin/jdb
+  fi
+fi
+ echo In set classpath
+if [ "$USE_GLOBAL_LOG4J" = true ]; then
```

Comment:
Should this be our default that needs to be switched off?

Reply:
If we change this to default it will impact on-prem customers on upgrade and they would have to change the flag to get the existing behavior and then migrate any logging config they have to the new file.   If we are willing to go that route then we can make this the default,  or we can change it to the default when we are happy with the direction internally and and are able to see it fine with customers.

Reply:
ok, let's bring it in and see where it gets us and we can change the default future forward.

Reply:
@wezell To note,  the setup of the tomcat scripts we override is a bit inconsistent.  Some of the core files we have updated and we also have our own versions,  this also gets copied to the separate bin and dotserver/tomcat.x.x.x/bin directories.   The preferred way to  customize tomcat is to have a separate TOMCAT_HOME for the immutable distribution distribution, and TOMCAT_BASE for any overrides.  Also CLASSPATH and JAVA_OPTS would be updated in a setenv.sh or setenv.bat.   We have customized startup.sh already which can cause problems and can be inconsistent.   startup.sh is only used when running in the background and running "./bin/catalina.sh run" is normal for running in the foreground e.g. in docker.  The classpath changes cannot be made in startup.sh as they are reset in catalina.sh before setenv.sh and setclasspath.sh are called anyway.   I would like to have our custom changes to JAVA_OPTS and CLASSPATH moved into a setenv.sh file but I do not know if we have left this for customers to be able to make their own customizations and we would cause problems by overriding this file if it already exists on upgrade.  I decided to put these changes into setclasspath.sh to be safe but is not a preferred approach.